File: Cpp/day-01.cpp
#include <algorithm> // For std::sort, std::count, and other algorithmss
#include <charconv>	 // For std::from_chars (fast string-to-int conversion)
#include <cstdint>	 // For int32_t
#include <cstdlib>	 // For std::abs (for integers)
#include <fstream>	 // For std::ifstream (file handling)
#include <iostream>	 // For std::cout, std::endl, std::cerr
#include <map>
#include <string> // For std::string, getline
#include <vector> // For std::vector (dynamic arrays)

// -- Placeholder for IDay.h ---
// Since we don't need the actual interface, we'll define a simple base class
// to satisfy the inheritance requirement without needing another file.
class IDay
{
};

// -- CLASS DEFINITION --
// Defines the main logic class, inheriting from the simple placeholder IDay.
class Day1 : public IDay
{
private:
	// Public members to store the data from the two columns. std::vector is the
	// C++ standard library's dynamic array
	std::vector<int32_t> list1;
	std::vector<int32_t> list2;

	// private function to handle all file I/O.
	bool ReadFileData();

public:
	// Declarations for the logic functions. The definitions follow below.
	void PartOne();
	void PartTwo();
};

bool Day1::ReadFileData()
{
	// Open the file. RAII: Destructor will automatically call myfile.close()
	// when scope ends.
	std::ifstream myfile("../inputs/input-01.txt.txt");

	if (!myfile.is_open())
	{
		std::cerr << "Error: Could not open file." << std::endl;
		return false;
	}

	int32_t left, right;

	// **IDIOMATIC C++ INPUT:** The `>>` operator reads formatted input.
	// It automatically skips all intermediate whitespace (spaces, tabs,
	// newlines) and attempts to convert the next non-whitespace sequence into
	// the target type (int32_t). The loop continues as long as two successful
	// reads occur.
	while (myfile >> left >> right)
	{
		list1.push_back(left);
		list2.push_back(right);
	}
	return true;
}

// --- METHOD IMPLEMENTATION: Day1::PartOne ---

void Day1::PartOne()
{
	// --- File Reading and Data Parsing ---

	if (!ReadFileData())
	{
		return; // Stop if file reading failed
	}

	// Sort the collected lists
	std::sort(list1.begin(), list1.end());
	std::sort(list2.begin(), list2.end());

	// Calculate Part One: total distance between lists
	long sum = 0;
	// Iterate and pair the sorted elements from both lists.
	for (size_t i = 0; i < list1.size(); i++)
		// Cast to 'long' before subtraction to avoid potential overflow during
		// calculation. std::abs from <cstdlib> is used for integer absolute
		// value.
		sum += std::abs((long)list1[i] - (long)list2[i]);

	std::cout << "Part 1: " << sum << std::endl;
}

void Day1::PartTwo()
{
	// OPTIMIZATION: Create a map to pre-count elements in list2 for faster
	// lookups. This makes the second loop much faster than repeatedly calling
	// std::count().
	std::map<int32_t, int> counts2;
	for (int32_t n : list2)
	{
		counts2[n]++;
	}

	long total = 0;
	// Calculate Part Two: similarity score
	for (int32_t i : list1)
	{
		// Multiply the element 'i' by its count found in the pre-calculated map
		// (counts2[i]). If i is not in the map, counts2[i] returns 0.
		total += (long)i * counts2[i];
	}

	std::cout << "Part 2: " << total << std::endl;
}

// --- MAIN ENTRY POINT ---

int main()
{
	// Create an instance of the solver class.
	Day1 solver;

	solver.PartOne();
	solver.PartTwo();

	return 0;
};

File: Go/day-01.go
package main

// Import necessary packages for file handling, string manipulation, formatting, and slices.
import (
	"fmt"     // Package for formatted I/O (like reading and printing)
	"os"      // Package for operating system functions (like reading files)
	"slices"  // Package for working with slices (used for sorting)
	"strings" // Package for string manipulation
)

// The main function is the entry point of the program.
func main() {
	// --- STEP 1: Reading the File and Error Handling ---

	// os.ReadFile("input.txt") reads the entire contents of the file into a byte slice.
	// The function returns two values: the data (byte slice) and an error.
	// The underscore (_) is used to ignore the potential error value for brevity.
	input, _ := os.ReadFile("../inputs/input-01.txt")

	// --- STEP 2: Initialization and Data Parsing ---

	// Declare two dynamic arrays (slices) to hold the numbers from the two columns.
	var list1, list2 []int

	// Declare a map (hash table) to store the count of each number in the second column (list2).
	// The keys are the numbers (int), and the values are their counts (int).
	counts2 := map[int]int{}

	// Convert the input byte slice to a string.
	inputStr := string(input)

	// strings.TrimSpace(inputStr) removes leading/trailing whitespace.
	// strings.Split(..., "\n") splits the string into a slice of strings based on newline characters,
	// giving us a slice where each element is a line from the file.
	for _, s := range strings.Split(strings.TrimSpace(inputStr), "\n") {
		// Declare two integer variables to hold the numbers read from the current line (s).
		var n1, n2 int

		// fmt.Sscanf reads formatted data from the string s and stores the results
		// into the memory addresses of n1 and n2.
		// The format "%d   %d" tells Sscanf to expect two integers separated by one or more spaces.
		// This is the Go equivalent of Python's mapping and splitting lines.
		fmt.Sscanf(s, "%d   %d", &n1, &n2)

		// append(list, element) dynamically adds the element to the slice.
		list1 = append(list1, n1)
		list2 = append(list2, n2)

		// Increment the count for the number n2 in the counts2 map.
		counts2[n2]++
	}

	// --- STEP 3: Sorting the Lists ---

	// slices.Sort performs an in-place sort on the elements of the slices.
	slices.Sort(list1)
	slices.Sort(list2)

	// --- STEP 4: Calculations ---

	// Initialize variables to store the results of the two calculations.
	part1, part2 := 0, 0

	// Iterate through the indices of list1 (since list1 and list2 are the same length).
	for i := range list1 {
		// --- Calculation for Part 1 (total distance between lists) ---
		// Note: Go's math.Abs only works on float64, so explicit casting is required.
		// The custom 'abs' function below is often preferred for integer math.
		part1 += abs(list2[i] - list1[i])

		// --- Calculation for Part 2 (similarity score) ---
		// It multiplies the current element of list1 by its count found in the second column data (counts2 map).
		part2 += list1[i] * counts2[list1[i]]
	}

	// --- STEP 5: Output Results ---

	fmt.Println(part1)
	fmt.Println(part2)
}

// Custom integer absolute value function.
// This is necessary because Go's built-in math.Abs requires float64 types.
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
